<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Catherine Benavides
- Ayudante: Nicolás Ojeda, Eduardo Moya

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Cristóbal Quilodran
- Nombre de alumno 2: Cristóbal Jara Cerda


### **Link de repositorio de GitHub:** `http://....`

### Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 7 días desde la publicación, 3 días de atraso con 1 punto de descuento c/u.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos Y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquer material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [32]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
!pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [33]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = '/content/drive/My Drive/lab2/'
except:
    print('Ignorando conexión drive-colab')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [34]:
df_grades = ...


<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **875** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **875** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier alumno que posea un valor diferente a esta escala, debería ser eliminado de la base de datos.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [35]:
#Cargamos la base de datos

#definimos antes los tipos de los atributos

dtypes={
    'names': str,
    'gender': str,
    'race/ethnicity': str,
    'parental level of education': str,
    'lunch': str,
    'test preparation course': str,
    'math score': int,
    'reading score': int,
    'writing score': int,
}

# Cargar el archivo JSON con Pandas
data = pd.read_json("/content/drive/MyDrive/lab3/students_grades.json", dtype=dtypes)

# Visualizar las primeras filas del DataFrame
display(data.head())

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course math score reading score writing score  
0  standard                    none         37            46            46  
1  standard                    none        124           142           142  
2  standard                    none        156           196           193  
3  standard               completed         80            78            81  
4  standard                    none        131           166           166

In [36]:
str_cols = ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']
for columna in str_cols:
    data[columna] = data[columna].replace('none', None)

# Convierte manualmente las columnas a los tipos de datos especificados
for columna, tipo in dtypes.items():
    if tipo == int:
        data[columna] = pd.to_numeric(data[columna], errors='coerce').astype('Int64')
    else:
        data[columna] = data[columna].astype(str)
data.dtypes

names                          object
gender                         object
race/ethnicity                 object
parental level of education    object
lunch                          object
test preparation course        object
math score                      Int64
reading score                   Int64
writing score                   Int64
dtype: object

In [37]:
#Creamos la función
def exploratory_data_analysis(dataframe):
   # Imprimir dimensiones del DataFrame
    print(f"1.- Dimensiones del DataFrame: {dataframe.shape}")
    print("------------------------------------------------------------------")

    # Imprimir nombres de las columnas
    print(f"2.- Nombres de las columnas: {dataframe.columns.tolist()}")
    print("------------------------------------------------------------------")

    # Mostrar los primeros 5, los últimos 5 elementos del DataFrame y un muestreo aleatorio de 5 elementos
    print("3.- a) Primeros 5 elementos del DataFrame:")
    display(dataframe.head())
    print("------------------------------------------------------------------")
    print("3.- b) Últimos 5 elementos del DataFrame:")
    display(dataframe.tail())
    print("------------------------------------------------------------------")
    print("3.- c) Muestreo aleatorio de 5 elementos del DataFrame:")
    display(dataframe.sample(5))
    print("------------------------------------------------------------------")

    # Descripción rápida de las variables numéricas
    print("4.- Descripción rápida de las variables numéricas:")
    display(dataframe.describe())
    print("------------------------------------------------------------------")

    # Cantidad de valores nulos
    print(f"5.- Cantidad de valores nulos por columna:\n{dataframe.isna().sum()}")
    print("------------------------------------------------------------------")

    # Cantidad de valores únicos por columna
    print(f"6.- Cantidad de valores únicos por columna:\n{dataframe.nunique()}")
    print("------------------------------------------------------------------")

    # Total de filas que contengan nombres duplicados
    duplicados = dataframe.duplicated(subset=["names"])
    duplicados_print = duplicados.sum()
    print(f"7.- Total de filas duplicadas: {duplicados_print}")
    print("------------------------------------------------------------------")

    # Filas que contengan nombres duplicados
    duplicated_data = dataframe[dataframe.loc[:, ['names']].duplicated()]
    print("8.- Filas que contienen nombres duplicados:")
    display(duplicated_data)
    pass

In [38]:
# Llamamos a la función
exploratory_data_analysis(data)

1.- Dimensiones del DataFrame: (1400, 9)
------------------------------------------------------------------
2.- Nombres de las columnas: ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']
------------------------------------------------------------------
3.- a) Primeros 5 elementos del DataFrame:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    None          37             46             46  
1  standard                    None         124            142            142  
2  standard                    None         156            196            193  
3  standard               completed          80             78             81  
4  standard                    None         131            166            166

------------------------------------------------------------------
3.- b) Últimos 5 elementos del DataFrame:


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1395      standard                    None         193            142   
1396  free/reduced               completed          66             74   
1397      standard               completed          54             61   
1398  free/reduced                    None         193            196   
1399      standard               completed          68             80   

      writing score  
1395            193  
1396             81  
1397             62  
1398            193  
1399             76

------------------------------------------------------------------
3.- c) Muestreo aleatorio de 5 elementos del DataFrame:


names  gender race/ethnicity parental level of education  \
745   Edward Killingsworth    male        group C                 high school   
889          Edward Fields    male        group D            some high school   
178           Cynthia Sisk  female        group D                some college   
1137        Anthony Pedigo    male        group B           bachelor's degree   
26           Samuel Dougal    male        group C             master's degree   

             lunch test preparation course  math score  reading score  \
745       standard                    None          70             70   
889       standard                    None          60             59   
178       standard                    None         193            171   
1137  free/reduced               completed          87             90   
26    free/reduced               completed          46             42   

      writing score  
745              65  
889              54  
178             193  
1137             88  
26               46

------------------------------------------------------------------
4.- Descripción rápida de las variables numéricas:


math score  reading score  writing score
count      1360.0         1360.0         1353.0
mean    89.538971      93.086029      91.852919
std     47.152422      46.385775       47.04357
min           0.0           17.0           10.0
25%          59.0           63.0           62.0
50%          72.0           76.0           75.0
75%         97.25          100.0          100.0
max         196.0          196.0          196.0

------------------------------------------------------------------
5.- Cantidad de valores nulos por columna:
names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     40
reading score                  40
writing score                  47
dtype: int64
------------------------------------------------------------------
6.- Cantidad de valores únicos por columna:
names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       96
reading score                    88
writing score                    92
dtype: int64
------------------------------------------------------------------
7.- Total de filas duplicadas: 247
---------------------

names  gender race/ethnicity parental level of education  \
12         Rita Courtney  female        group B            some high school   
81    Elizabeth Quintero  female        group B                some college   
160    Michelle Gonzales  female        group B            some high school   
170        Carla Mendoza  female        group D          associate's degree   
176       Darlene Parker  female        group C          associate's degree   
...                  ...     ...            ...                         ...   
1385      James Crawford    male        group D             master's degree   
1390     Gloria Bradford  female        group B                 high school   
1393         Adam Romero    male        group C                 high school   
1395    Alexander Dillon    male        group D          associate's degree   
1398        Joseph Mccoy    male        group D                some college   

             lunch test preparation course  math score  reading score  \
12        standard                    None         193            193   
81        standard                    None          82             85   
160       standard               completed          60             70   
170   free/reduced                    None          43             60   
176       standard                    None          54             61   
...            ...                     ...         ...            ...   
1385      standard                    None         193            147   
1390      standard                    None         124             64   
1393      standard               completed         124            193   
1395      standard                    None         193            142   
1398  free/reduced                    None         193            196   

      writing score  
12              193  
81               87  
160              74  
170              58  
176              58  
...             ...  
1385            131  
1390            193  
1393             49  
1395            193  
1398            193  

[247 rows x 9 columns]

In [39]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    data.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

**Justificación:**

```
9.- Es posible observar como anomalías que hay 1153 nombres distintos, más de los 875 supuestamente matrículados, que hay 247 estudiantes duplicados, y que hay valores nulos en las pruebas (aunque esto puede significar que tales estudiantes no rindieron esa prueba).

10.- Que hay valores sobre 100 (la escala es de 0 a 100) en los resultados de las pruebas, lo cual no es posible.
```

**Respuesta Esperada:**

In [ ]:
exploratory_data_analysis(df_grades)

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y cree una función que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**

- [ ] Crear función que permita limpiar el dataset.
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hint**

Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.


**Respuesta Esperada:**

In [41]:
def clean_data(data):
    # Eliminar duplicados basados en el nombre del estudiante
    data_sin_duplicados = data.drop_duplicates(subset=['names'])

    # Verificar que las notas estén dentro del rango válido (entre 0 y 100)
    notas_validas = ((data_sin_duplicados[['math score', 'reading score', 'writing score']] >= 0) &
                     (data_sin_duplicados[['math score', 'reading score', 'writing score']] <= 100)).all(axis=1)
    data_sin_duplicados = data_sin_duplicados[notas_validas]

    # Verificar que cada estudiante haya dado cada examen solo una vez
    examen_dado_una_vez = (data_sin_duplicados[['math score', 'reading score', 'writing score']].count(axis=1) == 3)
    data_sin_duplicados = data_sin_duplicados[examen_dado_una_vez]

    return data_sin_duplicados

In [42]:
df_grades = clean_data(data)
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    None          37             46   
3         standard               completed          80             78   
5         standard                    None          73             66   
6         standard               completed          81             75   
7         standard               completed          94             86   
...            ...                     ...         ...            ...   
1392      standard                    None          69             75   
1394  free/reduced               completed          23             44   
1396  free/reduced               completed          66             74   
1397      standard               completed          54             61   
1399      standard               completed          68             80   

      writing score  
0                46  
3                81  
5                63  
6                76  
7                87  
...             ...  
1392             71  
1394             36  
1396             81  
1397             62  
1399             76  

[875 rows x 9 columns]

> Utilice la función creada en el punto 1.1 para verificar sus resultados.

In [43]:
# Escriba su código aquí

def clean_data(data):
    # Eliminar duplicados basados en el nombre del estudiante
    data_sin_duplicados = data.drop_duplicates(subset=['names'])

    # Verificar que las notas estén dentro del rango válido (entre 0 y 100)
    notas_validas = ((data_sin_duplicados[['math score', 'reading score', 'writing score']] >= 0) &
                     (data_sin_duplicados[['math score', 'reading score', 'writing score']] <= 100)).all(axis=1)
    data_sin_duplicados = data_sin_duplicados[notas_validas]

    # Verificar que cada estudiante haya dado cada examen solo una vez
    examen_dado_una_vez = (data_sin_duplicados[['math score', 'reading score', 'writing score']].count(axis=1) == 3)
    data_sin_duplicados = data_sin_duplicados[examen_dado_una_vez]

    # Restablecer los índices
    data_sin_duplicados = data_sin_duplicados.reset_index(drop=True)

    return data_sin_duplicados

In [44]:
df_grades = clean_data(data)
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    None          37             46   
1        standard               completed          80             78   
2        standard                    None          73             66   
3        standard               completed          81             75   
4        standard               completed          94             86   
..            ...                     ...         ...            ...   
870      standard                    None          69             75   
871  free/reduced               completed          23             44   
872  free/reduced               completed          66             74   
873      standard               completed          54             61   
874      standard               completed          68             80   

     writing score  
0               46  
1               81  
2               63  
3               76  
4               87  
..             ...  
870             71  
871             36  
872             81  
873             62  
874             76  

[875 rows x 9 columns]

**Justificación:**

In [45]:
# Llamamos a la función del 1.1
exploratory_data_analysis(df_grades)

1.- Dimensiones del DataFrame: (875, 9)
------------------------------------------------------------------
2.- Nombres de las columnas: ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']
------------------------------------------------------------------
3.- a) Primeros 5 elementos del DataFrame:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    None          37             46             46  
1  standard               completed          80             78             81  
2  standard                    None          73             66             63  
3  standard               completed          81             75             76  
4  standard               completed          94             86             87

------------------------------------------------------------------
3.- b) Últimos 5 elementos del DataFrame:


names  gender race/ethnicity parental level of education  \
870   Richard Young    male        group D                 high school   
871   Wanda Russell  female        group B                 high school   
872  Marina Zeigler  female        group C           bachelor's degree   
873   Laurie Carter  female        group B            some high school   
874    Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
870      standard                    None          69             75   
871  free/reduced               completed          23             44   
872  free/reduced               completed          66             74   
873      standard               completed          54             61   
874      standard               completed          68             80   

     writing score  
870             71  
871             36  
872             81  
873             62  
874             76

------------------------------------------------------------------
3.- c) Muestreo aleatorio de 5 elementos del DataFrame:


names  gender race/ethnicity parental level of education  \
47   Donald Chestnut    male        group C                some college   
782     Oscar Markow    male        group C            some high school   
541       Julie Tang  female        group D          associate's degree   
278    Thomas Watson    male        group D                 high school   
753      Lorena Reed  female        group C                 high school   

            lunch test preparation course  math score  reading score  \
47       standard                    None          73             74   
782  free/reduced               completed          51             56   
541      standard                    None          85             91   
278  free/reduced               completed          73             68   
753  free/reduced                    None          36             53   

     writing score  
47              61  
782             53  
541             89  
278             66  
753             43

------------------------------------------------------------------
4.- Descripción rápida de las variables numéricas:


math score  reading score  writing score
count       875.0          875.0          875.0
mean    66.225143      69.195429      68.114286
std     15.076716       14.62927       15.17422
min           0.0           17.0           10.0
25%          57.0           59.0           57.5
50%          66.0           70.0           69.0
75%          76.5           79.0           79.0
max         100.0          100.0          100.0

------------------------------------------------------------------
5.- Cantidad de valores nulos por columna:
names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64
------------------------------------------------------------------
6.- Cantidad de valores únicos por columna:
names                          875
gender                           2
race/ethnicity                   5
parental level of education      6
lunch                            2
test preparation course          2
math score                      79
reading score                   72
writing score                   77
dtype: int64
------------------------------------------------------------------
7.- Total de filas duplicadas: 0
-----------------------------------------

Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []

```
Se puede observar que ya no existen anomalías y que el número de matriculados coincide con los 875 mencionados.
```

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [46]:
def nota_chilena(dataframe_in, columna):
    # Copiar el DataFrame de entrada para evitar modificar el original
    dataframe_out = dataframe_in.copy()

    # Aplicar la ecuación de la escala chilena a la columna especificada
    dataframe_out[columna] = (dataframe_in[columna] * 0.06) + 1

    # Redondear las notas transformadas al segundo decimal
    dataframe_out[columna] = dataframe_out[columna].round(2)

    return dataframe_out

In [47]:
df_grades= nota_chilena(df_grades, 'math score')
df_grades= nota_chilena(df_grades, 'reading score')
df_grades= nota_chilena(df_grades, 'writing score')
display(df_grades)

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    None        3.22           3.76   
1        standard               completed         5.8           5.68   
2        standard                    None        5.38           4.96   
3        standard               completed        5.86            5.5   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    None        5.14            5.5   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08            5.8   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
870           5.26  
871           3.16  
872           5.86  
873           4.72  
874           5.56  

[875 rows x 9 columns]

**Respuesta esperada:**

In [ ]:
df_grades

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [48]:
def alumnos_promedio(dataframe_in):
    # Copiar el DataFrame de entrada para evitar modificar el original
    dataframe_out = dataframe_in.copy()

    # Calcular el promedio de cada alumno
    promedio_alumno = dataframe_out[['math score', 'reading score', 'writing score']].mean(axis=1)

    # Redondear el promedio al segundo decimal
    promedio_alumno = promedio_alumno.round(2)

    # Agregar el promedio como una nueva columna al DataFrame original
    dataframe_out['GPA'] = promedio_alumno

    return dataframe_out

df_grades=alumnos_promedio(df_grades)
display(df_grades)

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    None        3.22           3.76   
1        standard               completed         5.8           5.68   
2        standard                    None        5.38           4.96   
3        standard               completed        5.86            5.5   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    None        5.14            5.5   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08            5.8   

     writing score   GPA  
0             3.76  3.58  
1             5.86  5.78  
2             4.78  5.04  
3             5.56  5.64  
4             6.22  6.34  
..             ...   ...  
870           5.26  5.30  
871           3.16  3.06  
872           5.86  5.42  
873           4.72  4.54  
874           5.56  5.48  

[875 rows x 10 columns]

**Respuesta Esperada:**

In [49]:
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    None        3.22           3.76           3.76   
1  standard               completed         5.8           5.68           5.86   
2  standard                    None        5.38           4.96           4.78   
3  standard               completed        5.86            5.5           5.56   
4  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
1  5.78  
2  5.04  
3  5.64  
4  6.34

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta esperada:**

In [ ]:
def reprobados_destacados(dataframe_in):
    pass

In [ ]:
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)
dataframe1

In [ ]:
dataframe2

In [ ]:
dataframe3

**Respuesta:**

In [50]:
def reprobados_destacados(dataframe_in):
    # Obtener DataFrame con los alumnos que reprobaron el año
    alumnos_reprobados = dataframe_in[dataframe_in['GPA'] < 4.0].reset_index(drop=True)

    # Obtener DataFrame similar al original, pero marcando con una "R" el atributo GPA de los alumnos reprobados
    dataframe_out = dataframe_in.copy()
    dataframe_out.loc[dataframe_out['GPA'] < 4.0, 'GPA'] = 'R'

    # Obtener DataFrame con los alumnos destacados ordenados de mayor a menor
    alumnos_destacados = dataframe_in[(dataframe_in['math score'] > 5.8) &
                                       (dataframe_in['reading score'] > 5.8) &
                                       (dataframe_in['writing score'] > 5.8)]
    alumnos_destacados = alumnos_destacados.sort_values(by='GPA', ascending=False).reset_index(drop=True)

    return alumnos_reprobados, dataframe_out, alumnos_destacados

In [51]:
dataframe1 , dataframe2, dataframe3 = reprobados_destacados(df_grades)

In [52]:
#Alumnos reprobados
dataframe1

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1       Mikki Amaya  female        group A                 high school   
2     Samuel Dougal    male        group C             master's degree   
3   Rodger Talavera    male        group B                 high school   
4    Morris Johnson    male        group A          associate's degree   
..              ...     ...            ...                         ...   
81   Charles Harian    male        group C            some high school   
82    Wendy Gunnell  female        group D            some high school   
83      Diana Costa  female        group C                 high school   
84      Amy Sieving  female        group C                 high school   
85    Wanda Russell  female        group B                 high school   

           lunch test preparation course  math score  reading score  \
0       standard                    None        3.22           3.76   
1   free/reduced               completed        3.04           3.88   
2   free/reduced               completed        3.76           3.52   
3   free/reduced                    None        3.16           2.74   
4   free/reduced               completed         3.4            4.3   
..           ...                     ...         ...            ...   
81  free/reduced               completed         3.7           4.12   
82  free/reduced                    None        2.62           3.04   
83  free/reduced                    None        3.46           3.76   
84      standard                    None        2.74           2.74   
85  free/reduced               completed        2.38           3.64   

    writing score   GPA  
0            3.76  3.58  
1            3.46  3.46  
2            3.76  3.68  
3            2.62  2.84  
4            4.18  3.96  
..            ...   ...  
81           3.94  3.92  
82           2.92  2.86  
83           3.58  3.60  
84            2.8  2.76  
85           3.16  3.06  

[86 rows x 10 columns]

In [53]:
#Total de alumnos pero reprobados marcados con R
dataframe2

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    None        3.22           3.76   
1        standard               completed         5.8           5.68   
2        standard                    None        5.38           4.96   
3        standard               completed        5.86            5.5   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    None        5.14            5.5   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08            5.8   

     writing score   GPA  
0             3.76     R  
1             5.86  5.78  
2             4.78  5.04  
3             5.56  5.64  
4             6.22  6.34  
..             ...   ...  
870           5.26   5.3  
871           3.16     R  
872           5.86  5.42  
873           4.72  4.54  
874           5.56  5.48  

[875 rows x 10 columns]

In [54]:
#Alumnos destacados de mejor a menor promedio
dataframe3

names  gender race/ethnicity parental level of education  \
0       Richard Pryor    male        group E           bachelor's degree   
1       Sandra Pompey  female        group E          associate's degree   
2   Elizabeth Beasley  female        group E           bachelor's degree   
3           Emma Gray  female        group E           bachelor's degree   
4        Emma Brasher  female        group D            some high school   
..                ...     ...            ...                         ...   
89     Rolando Widmer    male        group C                 high school   
90  Lucille Stanphill  female        group C                 high school   
91       Steven Kelly    male        group D                some college   
92        Peter Perez    male        group D             master's degree   
93        Harry Berry    male        group B          associate's degree   

       lunch test preparation course  math score  reading score  \
0   standard               completed         7.0            7.0   
1   standard                    None         7.0            7.0   
2   standard                    None         7.0            7.0   
3   standard               completed        6.94            7.0   
4   standard               completed        6.82            7.0   
..       ...                     ...         ...            ...   
89  standard               completed        5.92           6.04   
90  standard                    None        5.86           6.04   
91  standard                    None        5.86           5.92   
92  standard                    None        5.86           5.86   
93  standard               completed        5.86           5.92   

    writing score   GPA  
0             7.0  7.00  
1             7.0  7.00  
2             7.0  7.00  
3             7.0  6.98  
4             7.0  6.94  
..            ...   ...  
89           5.92  5.96  
90           5.92  5.94  
91           6.04  5.94  
92           6.04  5.92  
93           5.92  5.90  

[94 rows x 10 columns]

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [55]:
# Código para quintiles
def generar_quintiles(dataframe_in):
    # Generar quintiles de rendimiento para los alumnos según su promedio general
    quintiles = pd.qcut(dataframe_in['GPA'], q=5, labels=['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto'])

    # Agregar la columna de quintiles al DataFrame
    dataframe_in['GPA cuantiles'] = quintiles

    return dataframe_in

df_grades= generar_quintiles(df_grades)

In [56]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="GPA"),
    x="GPA",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig

```
La tendencia observada en el gráfico denota que a medida que los padres poseen un mayor nivel educacional, sus hijos poseen levemente un mayor GPA.
```

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

In [68]:
#Cargamos los archivos JSON con Pandas
data_1 = pd.read_json("/content/drive/MyDrive/lab3/students_grades_1.json", dtype=dtypes)
data_2 = pd.read_json("/content/drive/MyDrive/lab3/students_grades_2.json", dtype=dtypes)

In [69]:
# Concatenamos los DataFrames
data_unida = pd.concat([data_1, data_2], ignore_index=True)
exploratory_data_analysis(data_unida)

1.- Dimensiones del DataFrame: (875, 9)
------------------------------------------------------------------
2.- Nombres de las columnas: ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']
------------------------------------------------------------------
3.- a) Primeros 5 elementos del DataFrame:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none           3              3              3  
1  standard               completed           5              5              5  
2  standard                    none           5              4              4  
3  standard               completed           5              5              5  
4  standard               completed           6              6              6

------------------------------------------------------------------
3.- b) Últimos 5 elementos del DataFrame:


names  gender race/ethnicity parental level of education  \
870   Richard Young    male        group D                 high school   
871   Wanda Russell  female        group B                 high school   
872  Marina Zeigler  female        group C           bachelor's degree   
873   Laurie Carter  female        group B            some high school   
874    Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
870      standard                    none           5              5   
871  free/reduced               completed           2              3   
872  free/reduced               completed           4              5   
873      standard               completed           4              4   
874      standard               completed           5              5   

     writing score  
870              5  
871              3  
872              5  
873              4  
874              5

------------------------------------------------------------------
3.- c) Muestreo aleatorio de 5 elementos del DataFrame:


names  gender race/ethnicity parental level of education  \
116     Denise Chavez  female        group B                 high school   
321    Raymond Stuart    male        group A                some college   
679       Ronald Jett    male        group D                 high school   
551  Dolores Sprinkle  female        group C           bachelor's degree   
572   Larry Maldonado    male        group E             master's degree   

            lunch test preparation course  math score  reading score  \
116      standard               completed           5              5   
321      standard               completed           7              6   
679  free/reduced                    none           3              3   
551      standard                    none           5              5   
572      standard                    none           6              6   

     writing score  
116              5  
321              6  
679              3  
551              5  
572              6

------------------------------------------------------------------
4.- Descripción rápida de las variables numéricas:


math score  reading score  writing score
count  875.000000     875.000000     875.000000
mean     4.480000       4.672000       4.603429
std      0.946147       0.928117       0.943113
min      1.000000       2.000000       1.000000
25%      4.000000       4.000000       4.000000
50%      4.000000       5.000000       5.000000
75%      5.000000       5.000000       5.000000
max      7.000000       7.000000       7.000000

------------------------------------------------------------------
5.- Cantidad de valores nulos por columna:
names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64
------------------------------------------------------------------
6.- Cantidad de valores únicos por columna:
names                          875
gender                           2
race/ethnicity                   5
parental level of education      6
lunch                            2
test preparation course          2
math score                       7
reading score                    6
writing score                    7
dtype: int64
------------------------------------------------------------------
7.- Total de filas duplicadas: 0
-----------------------------------------

Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos] **


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


In [82]:
# Inserte su respuesta aquí

#Aplicamos la funcion de la 3.2 para obtener columna con promedios
data_unida = alumnos_promedio(data_unida)

# Definimos un diccionario para el posterior mapeo
mapeo_educacion = {
    'some high school': 'school',
    'some college': 'school',
    'high school': 'school',
    "bachelor's degree": 'college',
    "associate's degree": 'college',
    "master's degree": 'postgraduate'
}

# Aplicamos el diccionario con .map
data_unida['simple parental level of education'] = data_unida['parental level of education'].map(mapeo_educacion)

#Agregamos el promeido de notas según race/ethnicity y simple parental level of education
promedio_notas = data_unida.groupby(['race/ethnicity', 'simple parental level of education']).agg({'math score': 'mean', 'reading score': 'mean', 'writing score': 'mean', 'GPA': 'mean'}).reset_index()

# Agregamos el conteo de alumnos según race/ethnicity y simple parental level of education
conteo_alumnos = data_unida.groupby(['race/ethnicity', 'simple parental level of education']).size().reset_index(name='count')

# Calculamos el porcentaje respecto al total, lo hacemos string y le agregamos el '%'
total_alumnos = conteo_alumnos['count'].sum()
porcentaje_alumnos = (conteo_alumnos['count'] / total_alumnos) * 100
conteo_alumnos['percentage'] = porcentaje_alumnos.round(2).astype(str) + ' %'

# Fusionamos con merge el promedio de notas hecho con el de conteo
tabla_final = promedio_notas.merge(conteo_alumnos, on=['race/ethnicity', 'simple parental level of education'])
# Ordenamos con .sort
tabla_final = tabla_final.sort_values(by=['race/ethnicity', 'simple parental level of education']).reset_index(drop=True)

#Mostramos la tabla final
display(tabla_final)

race/ethnicity simple parental level of education  math score  \
0         group A                            college    4.125000   
1         group A                       postgraduate    4.500000   
2         group A                             school    4.117647   
3         group B                            college    4.518519   
4         group B                       postgraduate    4.400000   
5         group B                             school    4.196262   
6         group C                            college    4.490196   
7         group C                       postgraduate    4.466667   
8         group C                             school    4.296774   
9         group D                            college    4.614286   
10        group D                       postgraduate    4.750000   
11        group D                             school    4.530201   
12        group E                            college    5.057692   
13        group E                       postgraduate    5.000000   
14        group E                             school    4.920635   

    reading score  writing score       GPA  count percentage  
0        4.541667       4.375000  4.347917     24     2.74 %  
1        5.000000       5.000000  4.835000      2     0.23 %  
2        4.215686       4.058824  4.131176     51     5.83 %  
3        4.777778       4.685185  4.660370     54     6.17 %  
4        5.200000       5.000000  4.868000      5     0.57 %  
5        4.392523       4.271028  4.287009    107    12.23 %  
6        4.892157       4.882353  4.755882    102    11.66 %  
7        4.666667       4.733333  4.622667     15     1.71 %  
8        4.548387       4.432258  4.426129    155    17.71 %  
9        4.785714       4.757143  4.719571     70      8.0 %  
10       5.050000       5.300000  5.033500     20     2.29 %  
11       4.624161       4.617450  4.590537    149    17.03 %  
12       5.019231       5.000000  5.025769     52     5.94 %  
13       5.500000       5.500000  5.333333      6     0.69 %  
14       4.873016       4.682540  4.825714     63      7.2 %

In [84]:
fig = px.box(
    tabla_final,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

```
Sí, se observan diferenias en los GPA de acuerdo al nivel de educación de los padres y a la etnicidad, en concreto, la raza/etnia del grupo A presenta muy bajo GPA comparado al alto GPA de la raza/etnia del grupo E en los 3 niveles de educación de los padres.
```

**Resultado Esperado de la tabla**

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.74 |            5    |            4.89 |  4.88 |      24 | 2.74 %       |
|  1 |           | postgraduate                         |         4.69 |            5.23 |            5.35 |  5.09 |       2 | 0.23 %       |
|  2 |           | school                               |         4.57 |            4.73 |            4.56 |  4.62 |      51 | 5.83 %       |
|  3 | group B          | college                              |         5.07 |            5.26 |            5.19 |  5.18 |      54 | 6.17 %       |
|  4 |           | postgraduate                         |         4.91 |            5.69 |            5.55 |  5.38 |       5 | 0.57 %       |
|  5 |           | school                               |         4.69 |            4.89 |            4.76 |  4.78 |     107 | 12.23 %      |
|  6 | group C          | college                              |         5.02 |            5.37 |            5.35 |  5.25 |     102 | 11.66 %      |
|  7 |           | postgraduate                         |         4.92 |            5.14 |            5.1  |  5.06 |      15 | 1.71 %       |
|  8 |           | school                               |         4.76 |            5.02 |            4.92 |  4.9  |     155 | 17.71 %      |
|  9 | group D          | college                              |         5.11 |            5.25 |            5.25 |  5.2  |      70 | 8.0 %        |
| 10 |           | postgraduate                         |         5.22 |            5.54 |            5.73 |  5.5  |      20 | 2.29 %       |
| 11 |           | school                               |         5.02 |            5.13 |            5.11 |  5.09 |     149 | 17.03 %      |
| 12 | group E          | college                              |         5.54 |            5.45 |            5.45 |  5.48 |      52 | 5.94 %       |
| 13 |          | postgraduate                         |         5.54 |            6.03 |            5.89 |  5.82 |       6 | 0.69 %       |
| 14 |          | school                               |         5.4  |            5.31 |            5.16 |  5.29 |      63 | 7.2 %        |

### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene nuevamente los alumnos de los registros corruptos que ud. y su equipo filtraron en el análisis anterior.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

In [87]:
# Inserte su respuesta aquí

#Cargamos el nuevo archivo JSON con Pandas
data_otro = pd.read_csv("/content/drive/MyDrive/lab3/other_grades.csv", dtype=dtypes)

In [88]:
merged_outer = df_grades.merge(data_otro, how='outer')
print("Resultado del Outer Join:")
display(merged_outer)

Resultado del Outer Join:


names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
..                 ...     ...            ...                         ...   
995       Leona Markow     NaN            NaN                         NaN   
996  Maxine Gulbranson     NaN            NaN                         NaN   
997        Jeane Budds     NaN            NaN                         NaN   
998        Nancy Ramos     NaN            NaN                         NaN   
999     Andrea Mcdavid     NaN            NaN                         NaN   

        lunch test preparation course  math score  reading score  \
0    standard                    None        3.22           3.76   
1    standard               completed         5.8           5.68   
2    standard                    None        5.38           4.96   
3    standard               completed        5.86            5.5   
4    standard               completed        6.64           6.16   
..        ...                     ...         ...            ...   
995       NaN                     NaN        <NA>           <NA>   
996       NaN                     NaN        <NA>           <NA>   
997       NaN                     NaN        <NA>           <NA>   
998       NaN                     NaN        <NA>           <NA>   
999       NaN                     NaN        <NA>           <NA>   

     writing score   GPA GPA cuantiles  science score  history score  
0             3.76  3.58      Bajísimo             25            3.8  
1             5.86  5.78          Alto             56            5.2  
2             4.78  5.04         Medio             58            3.8  
3             5.56  5.64          Alto             72            5.0  
4             6.22  6.34      Muy alto             94            4.6  
..             ...   ...           ...            ...            ...  
995           <NA>   NaN           NaN             39            3.8  
996           <NA>   NaN           NaN             38            3.9  
997           <NA>   NaN           NaN             43            3.9  
998           <NA>   NaN           NaN             54            6.2  
999           <NA>   NaN           NaN             40            4.2  

[1000 rows x 13 columns]

In [89]:
merged_left = df_grades.merge(data_otro, how='left')
print("Resultado del Left Join:")
display(merged_left)

Resultado del Left Join:


names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    None        3.22           3.76   
1        standard               completed         5.8           5.68   
2        standard                    None        5.38           4.96   
3        standard               completed        5.86            5.5   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    None        5.14            5.5   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08            5.8   

     writing score   GPA GPA cuantiles  science score  history score  
0             3.76  3.58      Bajísimo             25            3.8  
1             5.86  5.78          Alto             56            5.2  
2             4.78  5.04         Medio             58            3.8  
3             5.56  5.64          Alto             72            5.0  
4             6.22  6.34      Muy alto             94            4.6  
..             ...   ...           ...            ...            ...  
870           5.26  5.30         Medio             62            4.2  
871           3.16  3.06      Bajísimo             16            3.6  
872           5.86  5.42          Alto             59            4.1  
873           4.72  4.54          Bajo             43            4.3  
874           5.56  5.48          Alto             61            5.3  

[875 rows x 13 columns]

In [90]:
merged_right = df_grades.merge(data_otro, how='right')
print("Resultado del Right Join:")
display(merged_right)

Resultado del Right Join:


names  gender race/ethnicity parental level of education  \
0           Pam Otoole  female        group B           bachelor's degree   
1          Diane Olson  female        group C                some college   
2       Sarah Flannery  female        group B             master's degree   
3     Stanley Chappell    male        group A          associate's degree   
4        Arnold Dodson    male        group C                some college   
..                 ...     ...            ...                         ...   
995    Melanie Acevedo  female        group E             master's degree   
996   Christopher Wall    male        group C                 high school   
997  Margaret Spannaus  female        group C                 high school   
998     Vickie Sellers  female        group D                some college   
999      Kari Mitchell  female        group D                some college   

            lunch test preparation course  math score  reading score  \
0        standard                    None        5.32           5.32   
1        standard               completed        5.14            6.4   
2        standard                    None         6.4            6.7   
3    free/reduced                    None        3.82           4.42   
4        standard                    None        5.56           5.68   
..            ...                     ...         ...            ...   
995      standard               completed        6.28           6.94   
996  free/reduced                    None        4.72            4.3   
997  free/reduced               completed        4.54           5.26   
998      standard               completed        5.08           5.68   
999  free/reduced                    None        5.62           6.16   

     writing score   GPA GPA cuantiles  science score  history score  
0             5.44  5.36          Alto             57            4.0  
1             6.28  5.94      Muy alto             48            4.8  
2             6.58  6.56      Muy alto             72            6.1  
3             3.64  3.96      Bajísimo             47            3.7  
4              5.5  5.58          Alto             60            5.2  
..             ...   ...           ...            ...            ...  
995            6.7  6.64      Muy alto             79            6.3  
996            4.3  4.44          Bajo             62            4.0  
997            4.9  4.90          Bajo             47            4.0  
998           5.62  5.46          Alto             47            5.7  
999           6.16  5.98      Muy alto             53            6.2  

[1000 rows x 13 columns]

In [91]:
merged_inner = df_grades.merge(data_otro, how='inner')
print("Resultado del Inner Join:")
display(merged_inner)

Resultado del Inner Join:


names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    None        3.22           3.76   
1        standard               completed         5.8           5.68   
2        standard                    None        5.38           4.96   
3        standard               completed        5.86            5.5   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    None        5.14            5.5   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08            5.8   

     writing score   GPA GPA cuantiles  science score  history score  
0             3.76  3.58      Bajísimo             25            3.8  
1             5.86  5.78          Alto             56            5.2  
2             4.78  5.04         Medio             58            3.8  
3             5.56  5.64          Alto             72            5.0  
4             6.22  6.34      Muy alto             94            4.6  
..             ...   ...           ...            ...            ...  
870           5.26  5.30         Medio             62            4.2  
871           3.16  3.06      Bajísimo             16            3.6  
872           5.86  5.42          Alto             59            4.1  
873           4.72  4.54          Bajo             43            4.3  
874           5.56  5.48          Alto             61            5.3  

[875 rows x 13 columns]

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>